tcms/testcases/views.py

```diff
@@ -58,74 +58,78 @@ def plan_from_request_or_none(request, pk_enough=False):
     Arguments:
     - pk_enough: a choice for invoker to determine whether the ID is enough.
     """
-    tp_id = request.POST.get("from_plan") or request.GET.get("from_plan")
-    if tp_id:
+    test_plan_id = request.POST.get("from_plan") or request.GET.get("from_plan")
+    if test_plan_id:
         if pk_enough:
             try:
-                tp = int(tp_id)
+                test_plan = int(test_plan_id)
             except ValueError:
-                tp = None
+                test_plan = None
         else:
-            tp = get_object_or_404(TestPlan, plan_id=tp_id)
+            test_plan = get_object_or_404(TestPlan, plan_id=test_plan_id)
     else:
-        tp = None
-    return tp
+        test_plan = None
+    return test_plan
 
 
-def update_case_email_settings(tc, n_form):
+def update_case_email_settings(test_case, n_form):
     """Update testcase's email settings."""
 
-    tc.emailing.notify_on_case_update = n_form.cleaned_data[
+    test_case.emailing.notify_on_case_update = n_form.cleaned_data[
         'notify_on_case_update']
-    tc.emailing.notify_on_case_delete = n_form.cleaned_data[
+    test_case.emailing.notify_on_case_delete = n_form.cleaned_data[
         'notify_on_case_delete']
-    tc.emailing.auto_to_case_author = n_form.cleaned_data[
+    test_case.emailing.auto_to_case_author = n_form.cleaned_data[
         'author']
-    tc.emailing.auto_to_case_tester = n_form.cleaned_data[
+    test_case.emailing.auto_to_case_tester = n_form.cleaned_data[
         'default_tester_of_case']
-    tc.emailing.auto_to_run_manager = n_form.cleaned_data[
+    test_case.emailing.auto_to_run_manager = n_form.cleaned_data[
         'managers_of_runs']
-    tc.emailing.auto_to_run_tester = n_form.cleaned_data[
+    test_case.emailing.auto_to_run_tester = n_form.cleaned_data[
         'default_testers_of_runs']
-    tc.emailing.auto_to_case_run_assignee = n_form.cleaned_data[
+    test_case.emailing.auto_to_case_run_assignee = n_form.cleaned_data[
         'assignees_of_case_runs']
 
     default_tester = n_form.cleaned_data['default_tester_of_case']
-    if (default_tester and tc.default_tester_id):
-        tc.emailing.auto_to_case_tester = True
+    if (default_tester and test_case.default_tester_id):
+        test_case.emailing.auto_to_case_tester = True
 
     # Continue to update CC list
     valid_emails = n_form.cleaned_data['cc_list']
-    tc.emailing.cc_list = CC_LIST_DEFAULT_DELIMITER.join(valid_emails)
+    test_case.emailing.cc_list = CC_LIST_DEFAULT_DELIMITER.join(valid_emails)
 
-    tc.emailing.save()
+    test_case.emailing.save()
 
 
 def group_case_bugs(bugs):
     """Group bugs using bug_id."""
     bugs = itertools.groupby(bugs, lambda b: b.bug_id)
-    bugs = [(pk, list(_bugs)) for pk, _bugs in bugs]
+    bugs = []
+
+    for key, _bugs in bugs:
+        bugs.append((key, list(_bugs)))
+
```

Comment:
This is breaking existing behavior, so you will have to add a test for that, and also fix it. 

The problem lies in how list comprehension is evaluated. First the right part is evaluated, where `bugs` is the input parameter, which is an iterator (coming from line 106). The data is constructed into a list and assigned again to the `bugs` variable, changing its data type.

In your implementation L106 `bugs` is an iterator. Then L107 makes `bugs` an empty list. Then L109 executes a loop over the empty list with name `bugs`, which doesn't go into any iterations so on L112 you remove an empty result when the input parameter wasn't empty.

Reply:
Note: the entire queryset to iterator (with group by that can be done via QS) to a list of tuples, containing list of bugs is beyond me. Let's fix the obvious issue first (with a test case) and later work on refining this function.

Note2: Move this change into another PR and ping me to get some help about writing a test for this. Once that is merged we can return here and rebase.

Reply:
a variable called bugs is causing a bug lol